### ISSUE TITLE
Autocomplete in Array and Dictionary Index Operator does not work correctly

### ISSUE THREAD

**Issue by zolrath** [godotengine/godot#83074]
### Godot version

4.2.dev6

### System information

Windows 11

### Issue description

When using autocomplete to index an array or dictionary it does not list variable names for autocomplete, instead it lists the available methods as if I had typed a `.`
![image](https://github.com/godotengine/godot/assets/454563/d17b98c2-b21c-4256-84dd-0098b4ab7298)

### Steps to reproduce

1. Create a variable containing an index: `var index = 0`
2. Create an array or a dictionary, `var test = [1,2,3]`
3. Attempt to use autocomplete to index the array using the index variable. `test[in]`
4. Note that the autocomplete results have the same methods as putting a `.` instead of providing relevant autocomplete variables.
![image](https://github.com/godotengine/godot/assets/454563/2dd3e06a-f271-429e-b153-25a51d2b3443)
![image](https://github.com/godotengine/godot/assets/454563/37c41681-ff26-44b9-86f2-261392d1023a)

### Minimal reproduction project

N/A

**Comment by GrammAcc** [comment]
Hi, I'm looking for my next contribution, and I'd like to take a shot at fixing this if no one else is working on it.

I'm on my lunch break right now, so I can't dig into the code, but I suspect it has something to do with the fact that Dictionary keys can be accessed via property access in GDScript, so I will start by trying to pin down the area of the code that is responsible for code completion in containers and see if I can find a correlation with the property access of container contents.

Also, it seems odd that we would have autocomplete for Array contents anyway since the contents of an Array are not identifiers, and the only situation where we wouldn't just use a literal int when using item access on Arrays is when accessing items by dynamic index, in which case, the index would be stored in an identifier in the current scope like in the example provided by the OP.

We should probably have auto-completion for literal Dictionary keys when using item access, but I think it's reasonable to simply use the current scope for auto-complete suggestions when using item access on an Array. Also, including identifiers from the current scope in the suggestions for Dictionary keys during item access would likely be a benefit as well since a common pattern I've used in the past is to access Dictionary values dynamically with the keys stored in a variable. Thinking about it, this might be a bad practice, but it's often a really simple way to handle data processing in a loop for things like loading a saved game.

I also tried accessing the contents of a Dictionary with property access, and they don't show up in the suggestions, so a fix should also make sure that Dictionary keys are included in suggestions when using property access.

![screen-1_shot_2023-10-10_11:22:18](https://github.com/godotengine/godot/assets/35282898/bc88c39c-2d2f-4201-8638-3e8e13af38e8)

One other consideration that comes to mind right now is how this might affect things like Enums. I often use constant Enums in my projects for things like movement states and whatnot, and the autocomplete suggestions are really helpful for long ALL_CAPS enum members, so if this is an issue with containers in general, I will need to make sure the changes don't affect more specialized containers like Enums.

If anyone with more knowledge of this area has a better idea of how to handle the suggestions for containers, please let me know. Thanks!

I'll start investigating the source of the incorrect suggestions after work.

**Comment by zolrath** [comment]
Hm, with the dictionary example you included the property access does work if the dictionary was defined within the function you're accessing it in:

![image](https://github.com/godotengine/godot/assets/454563/094b864b-630a-4134-b228-c1bc95324e7d)

But not if defined outside of that scope:

![image](https://github.com/godotengine/godot/assets/454563/77b88a61-63cc-43cf-89f5-6247f91d6133)

**Comment by GrammAcc** [comment]
I poked around in the source code for the GDScript parser, but I'm going to need to dig around some more to familiarize myself with how the completions are being generated.

I'll keep working on this, but I probably won't be able to get a PR put together until around the 21st. Saturday is my only day off each week, and I'm going to a local coding event this weekend, so I won't be able to do much this Saturday.

I'm guessing this isn't urgent since it's mostly QOL, but if this needs a faster fix, I can pass on whatever I have so far to someone with more time. Just let me know. :)

**Comment by GrammAcc** [comment]
Just an FYI that I'm still working on this one, but it's taking a while to figure out what's going on.

When generating completion suggestions, the editor parses available symbols block by block, and the function block should have a reference to a parent block pointing to the class scope, but for some reason, that parent block is null when the completions are parsed in function scope, so it's not picking up any variables declared in class scope.

I spent several hours digging around the gdscript module of the source code, but this module is pretty hard to navigate since almost every relevant function mutates pass-by-reference arguments and some of them directly mutate relevant state on the parser class itself, so figuring out exactly where the parent block is getting nulled out is proving difficult.

On the bright side, I think I understand how the editor is actually putting together the completion suggestions and how it decides which suggestions to include for attribute vs item access, so once I find the cause of the scope problems, I should be able to get a fix put together fairly quickly \*knock on wood.

I'll probably have to get some help in the contributors chat on this one, but I'm out of time for today, so it will have to wait.
I'll keep working on a PR, but I don't know when I will finally get something put together.

**Comment by HolonProduction** [comment]
I guess this is a duplicate of #76955 and #79378 would fix it.

**Comment by HolonProduction** [comment]
To explain what's going on. In GDScript you can access the variables of a class by using the `instance["name"]` syntax. For that reason writing `instance[]` will result in the same completion context as `instance.` (Completion type subscript). Therefore the auto completion did a method lookup. To mitigate this we can check whether the type on which the subscript happens is a builtin type (like array or dict, other builtin types do not support the [] syntax to access variables). In that case we do a identifier lookup instead of looking up the methods.

**Comment by akien-mga** [comment]
- Duplicate of https://github.com/godotengine/godot/issues/76955

**Comment by zolrath** [comment]
@HolonProduction Would that PR also correct the dictionary behavior changing because of scope posted in 
 https://github.com/godotengine/godot/issues/83074#issuecomment-1756065087or should I split that into its own ticket?

**Comment by HolonProduction** [comment]
> @HolonProduction Would that PR also correct the dictionary behavior changing because of scope posted in [#83074 (comment)](https://github.com/godotengine/godot/issues/83074#issuecomment-1756065087) or should I split that into its own ticket?

Good point. My current PR would break suggestion for dictionary keys completely.

When it comes to the scope inconsistency there is already an issue for it #80145. As mentioned there, i think this is the correct behavior.
